### ISSUE TITLE
Set Standardized Procedures for Issue Management

### ISSUE THREAD

**Issue by wpank** [sourcecred/sourcecred#1064]
#### "At the point of submission of an issue, the user likely has little patience or interest in providing expansive detail. As such, you should make it as easy as possible to get the most useful information from them in the shortest time possible" - Jono Bacon
In an effort to make Sourcecred more open source friendly, creating good practices around Issues helps create a transparent framework for community collaboration. 

The following proposes some standards and procedures to help streamline the process of managing Issues. 

## 1.) Make an Issue Template
Issue templates streamline the process for contributors to add new issues. For a new contributor this saves the hassle of trying to find out what format to use or what items to include. 

[Github describes how to make an issue template for a repository here.](https://help.github.com/articles/manually-creating-a-single-issue-template-for-your-repository/)

An issue template should provide a boilerplate structure for necessary sections of an issue. 

Possible sections:

* `Context around the issue`
* `Checklist of items that will close the issue`
* `Implementation Plan`
* `Test Plan`
* `Estimated time/resources required`

Examples:

* [Github Issue Templates](https://github.com/stevemao/github-issue-templates)
* [Awesome Github Templates](https://github.com/devspace/awesome-github-templates)

## 2.) Add a section to Contributing on how to write a good issue
Outlining how to write quality issues leads to better open source. Well defined issues lower the barrier to entry for a contributor to jump in and also reduces potential back and forth discussion. Defining these guidelines in Contributing helps newcomers get a good foundation of practices on how to make quality issues in the future (which benefits all projects :grinning:)

Potential suggestions:

* Define the issue as clearly as possible. Someone unfamiliar with the project should be able to  understand it
* The issue needs to provide something of value for the project
* The issue is independent of other issues
* Small and easily estimable in terms of time and resources required
* Measureable / quantifiable: the should be a clear end result
* Describes the context around how and why the issue is needed. Link to other issues to further define a task
* Make a list of actionable items that need to be completed in order for the issue to be closed
* Describe the issue in a way that it falls under a certain category. ie `bug`, `feature`, `documentation`, `refactor`, or whatever paradigm the project follows. This helps with triaging issues. 
* Mention any issues that an issue may block, or if the issue is blocked by other issues

### Label Categories
The current labels in sourcecred only describe a limited subset of information at a glance. Is this issue really important? Is someone  already in progress with it? Following consistent labeling schemes that include this sort of information helps keep the state of the project transparent and explicitly clear. 

Labels can also be a source of signal for cred, where making a pull request for a `High Priority` label gives more cred then submitting a pull request for a `Low Priority` label. 

#### Proposed Labeling Scheme
I like what Dave Lunny describes [Here](https://medium.com/@dave_lunny/sane-github-labels-c5d2e6004b63). 

Labels should have three categories:

* Type
* Priority
* Status

##### Type
This is more or less what sourcecred's labels describe at the moment. 
The article suggests:

* `Type: Bug`
* `Type: Enhancement`
* `Type: Maintenence`
* `Type: Question`

#### Priority
Describing the priority of an issue helps triage issues and designate where the focus lies with development. At the moment sourcecred's issues don't mention this much. 
The article suggests:

* `Priority: Critical`
* `Priority: High`
* `Priority: Medium`
* `Priority: Low`

#### Status
Describing where the issues is in it's state of development can help with determining where to allocate resources. If an issue has nothing but a description, someone can then move on fleshing out an implementation instead of wondering if someone already has a branch open for it. 
The article suggests:

* `Status: Abandoned`
* `Status: Accepted`
* `Status: Available`
* `Status: Blocked`
* `Status: Completed`
* `Status: In Progress`
* `Status: On Hold`
* `Status: Pending`
* `Status: Review Needed`
* `Status: Revision Needed`

The author of that article has a cool command line tool for making git labels: [git-labelmaker](https://github.com/himynameisdave/git-labelmaker)

#### [Stale](https://probot.github.io/apps/stale/)
Stale is a bot that will label an issue as stale if no activity has happened after a certain period of time. This can help contributor's determine what work is currently being done (or what someone has forgotten about and may be useful to work on). It can also help with triaging issues in terms of priority (stale issues may turn out to be important). 

#### [Triage New Issues](https://probot.github.io/apps/triage-new-issues/)
Triage New Issues is a bot that will label a new issue with a `triage` label. This can create a streamlined process of labeling, whereby a triager has clearly defined issues to assess. This then avoids the confusion of "has this issue without a tag been addressed?". 

#### [Request Info](https://probot.github.io/apps/request-info/)
Request Info is a bot that will request more info from a newly created Issue or Pull Request that is left blank. This can be useful for contributors that assume describing something in the title of an issue or pull request is sufficient. It removes the need to ask for more information from a contributor/maintainer. 

#### [helPR](https://probot.github.io/apps/helpr/)
helPR is a bot that will make labels for when an issue has a PR availalbe. It will apply the following labels depending on the state of a PR: `pr-available`, `pr-merged`, and `pr-rejected`. This helps with knowing whether work is currently being done on an issue. If an issue has the label `pr-available`, it could be a better use of my time to look at an issue that doesn't have any PR's yet. 

#### [Issue Complete](https://probot.github.io/apps/issue-complete/)
Issue Complete is a bot that will look at the body of an issue and add a label and comment if it has unchecked boxes in the tasklist and/or has missing keywords. An example use case would be to have an issue template with 

* [x] I have sufficiently described the issue and context around the issue
* [x] I have provided a sufficient test case describing the intended outcome

where the user will check both of these and if not, the bot will complain. 

#### [Create a Github Issue Triage Bot](https://github.blog/2018-10-31-automating-issue-triage-with-github-and-recastai/)
The article linked is a github post that describes how to make a Github App that will react to actions on github. This could be a fun project for getting familiar with Github's webhooks api. 

## Further Resources / References
[Github Issue Triage and Transparency](https://medium.com/@clarkbw/github-issue-triage-and-transparency-be35acd8e85d)
[How We Use Github Issues To Organize a Project](http://www.ianbicking.org/blog/2014/03/use-github-issues-to-organize-a-project.html)
[How To Be An Open Source Gardener](https://words.steveklabnik.com/how-to-be-an-open-source-gardener)

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200032326400385/1200038364142464) by [Unito](https://www.unito.io/learn-more)

**Comment by wpank** [comment]
> To check my understanding: not all of the “potential suggestions” for
issues should be requirements, right? Being small and requiring
predictable resource allocation, and being measurable or quantifiable,
don’t seem like they always apply to things that are otherwise valuable
issues: SourceCred’s #476, #862, #913, and #945 are some examples. If
not issues, where should this information go?

I would say that they're more suggestions. If an issue can be broken down and framed in those manners, vs not being framed like that, I think it helps for causal contributors to know the scope around what it may require for them to take that issue on. 

I like to question it as: is an issue bounty-able? If someone can determine the amount of money it takes to solve an issue, it is framed nicely from a project management perspective. 

Of course not all issues can be presented this way, where issues like #476, #862, #913, and #945 require some discussion to figure out scope. In this case they should be presented as they are. 

> Want to quickly note that while I don’t mind stalebot labeling issues
as stale, I absolutely abhor closebot and its even more sinister cousin
lockbot. The effect of these bots is to make sure that you do not have
a single source of truth for a single issue, that people are forced to
create duplicate issues ad nauseam, and that it is harder to find
issues that have not been resolved. So, barring good arguments to the
contrary, I’m opposed to their use in SourceCred.

I would say the goal here would be to add non-obtrusive bots that help with maintenance but don't require more maintenance. These might be more useful for much larger projects that move faster. I haven't used them much myself, so I can't comment on their usefulness.
